cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java

```diff
@@ -544,25 +549,98 @@ private boolean needsReconfiguring(Reconciliation reconciliation, String connect
             boolean shouldPause = Boolean.TRUE.equals(connectorSpec.getPause());
             if ("RUNNING".equals(state) && shouldPause) {
                 log.debug("{}: Pausing connector {}", reconciliation, connectorName);
-                return apiClient.pause(host, port,
-                        connectorName)
-                        .compose(ignored ->
-                                apiClient.status(host, port,
-                                        connectorName));
+                return apiClient.pause(host, port, connectorName);
             } else if ("PAUSED".equals(state) && !shouldPause) {
                 log.debug("{}: Resuming connector {}", reconciliation, connectorName);
-                return apiClient.resume(host, port,
-                        connectorName)
-                        .compose(ignored ->
-                                apiClient.status(host, port,
-                                        connectorName));
-
+                return apiClient.resume(host, port, connectorName);
             } else {
-                return Future.succeededFuture(status);
+                return Future.succeededFuture();
             }
         }
     }
 
+    private Future<Void> maybeRestartConnector(Reconciliation reconciliation, String host, KafkaConnectApi apiClient, String connectorName, CustomResource resource) {
+        if (hasRestartAnnotation(resource, connectorName)) {
+            log.debug("{}: Restarting connector {}", reconciliation, connectorName);
+            return apiClient.restart(host, port, connectorName)
+                    .compose(ignored -> removeRestartAnnotation(reconciliation, resource),
+                        throwable -> {
+                            // Ignore restart failures - just try again on the next reconcile
+                            log.warn("{}: Failed to restart connector {}. {}", reconciliation, connectorName, throwable.getMessage());
+                            return Future.succeededFuture();
+                        });
+        } else {
+            return Future.succeededFuture();
+        }
+    }
+
+    private Future<Void> maybeRestartConnectorTask(Reconciliation reconciliation, String host, KafkaConnectApi apiClient, String connectorName, CustomResource resource) {
+        int taskID = getRestartTaskAnnotationTaskID(resource, connectorName);
+        if (taskID >= 0) {
+            log.debug("{}: Restarting connector task {}:{}", reconciliation, connectorName, taskID);
+            return apiClient.restartTask(host, port, connectorName, taskID)
+                    .compose(ignored -> removeRestartTaskAnnotation(reconciliation, resource),
+                        throwable -> {
+                            // Ignore restart failures - just try again on the next reconcile
+                            log.warn("{}: Failed to restart connector task {}:{}. {}", reconciliation, connectorName, taskID, throwable.getMessage());
```

Comment:
I have my doubts about this, particularly in the case that the connector task doesn't exist (i.e. bad id, presumably an http 404). If it's only in the logs then the user might not realise why the annotation wasn't removed. For transient errors it will likely succeed, so leaving it in the annotation for another try is not a bad idea. But for permanent errors like non-existence its more of a problem, especially if the user doesn't have access to the logs. I think we should propagate the error as a warning in the status. 

The error case doesn't seem to be covered by the tests.

Reply:
The latest commit adds warning status conditions to the CR when the restart or restartTask REST calls fail. Have also added failure test cases.